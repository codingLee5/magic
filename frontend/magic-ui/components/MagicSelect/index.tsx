import { IconChevronDown } from "@tabler/icons-react"
import type { SelectProps } from "antd"
import { Select } from "antd"
import MagicIcon from "../MagicIcon"
import { memo } from "react"

export type MagicSelectProps = SelectProps

const suffixIcon = <MagicIcon component={IconChevronDown} size={16} />

const MagicSelect = memo(({ ...props }: MagicSelectProps) => {
	return <Select suffixIcon={suffixIcon} {...props} />
})

MagicSelect.displayName = "MagicSelect"

export default MagicSelect
