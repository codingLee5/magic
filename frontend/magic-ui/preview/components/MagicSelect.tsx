import React from "react"
import { Space } from "antd"
import MagicSelect from "../../components/MagicSelect"
import ComponentDemo from "./Container"

const SelectDemo: React.FC = () => {
	const options = [
		{ value: "option1", label: "选项1" },
		{ value: "option2", label: "选项2" },
		{ value: "option3", label: "选项3" },
		{ value: "option4", label: "选项4" },
		{ value: "option5", label: "选项5" },
	]

	return (
		<div>
			<ComponentDemo
				title="基础选择器"
				description="最基本的选择器组件"
				code="<MagicSelect placeholder='请选择' options={options} />"
			>
				<Space>
					<MagicSelect placeholder="请选择" style={{ width: 200 }} options={options} />
				</Space>
			</ComponentDemo>

			<ComponentDemo title="多选选择器" description="支持多选的选择器" code="mode='multiple'">
				<Space>
					<MagicSelect
						mode="multiple"
						placeholder="请选择多个选项"
						style={{ width: 300 }}
						options={options}
					/>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="不同尺寸"
				description="支持不同尺寸的选择器"
				code="size: 'large' | 'middle' | 'small'"
			>
				<Space direction="vertical">
					<MagicSelect
						size="large"
						placeholder="大尺寸选择器"
						style={{ width: 200 }}
						options={options}
					/>
					<MagicSelect
						placeholder="默认尺寸选择器"
						style={{ width: 200 }}
						options={options}
					/>
					<MagicSelect
						size="small"
						placeholder="小尺寸选择器"
						style={{ width: 200 }}
						options={options}
					/>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="可搜索选择器"
				description="支持搜索功能的选择器"
				code="showSearch"
			>
				<Space>
					<MagicSelect
						showSearch
						placeholder="可搜索的选择器"
						style={{ width: 200 }}
						options={options}
						filterOption={(input, option) =>
							!!option?.value?.toString().includes(input)
						}
					/>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="事件处理"
				description="监听选择变化事件"
				code="onChange | onSelect | onDeselect"
			>
				<Space direction="vertical">
					<MagicSelect
						placeholder="监听选择变化"
						style={{ width: 200 }}
						options={options}
						onChange={(value) => console.log("选择的值:", value)}
					/>
					<MagicSelect
						mode="multiple"
						placeholder="监听多选变化"
						style={{ width: 300 }}
						options={options}
						onChange={(value) => console.log("多选的值:", value)}
					/>
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default SelectDemo
