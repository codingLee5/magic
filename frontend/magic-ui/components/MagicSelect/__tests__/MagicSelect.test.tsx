import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicSelect from "../index"

describe("MagicSelect", () => {
	it("should render without crashing", () => {
		const options = [
			{ label: "选项1", value: "1" },
			{ label: "选项2", value: "2" },
			{ label: "选项3", value: "3" },
		]

		render(
			<MagicThemeProvider theme="light">
				<MagicSelect options={options} placeholder="请选择" />
			</MagicThemeProvider>,
		)

		// 检查是否渲染了占位符
		expect(screen.getByText("请选择")).toBeInTheDocument()
	})

	it("should display selected value", () => {
		const options = [
			{ label: "选项1", value: "1" },
			{ label: "选项2", value: "2" },
		]

		render(
			<MagicThemeProvider theme="light">
				<MagicSelect options={options} value="1" />
			</MagicThemeProvider>,
		)

		// 检查是否显示了选中的值
		expect(screen.getByText("选项1")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
				{ label: "选项3", value: "3" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} placeholder="请选择" />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带默认值选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} defaultValue="1" />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("禁用状态选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} disabled />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("多选选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
				{ label: "选项3", value: "3" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} mode="multiple" />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("大尺寸选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} size="large" />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("小尺寸选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} size="small" />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带清除按钮选择器快照", () => {
			const options = [
				{ label: "选项1", value: "1" },
				{ label: "选项2", value: "2" },
			]

			const { asFragment } = render(
				<MagicThemeProvider theme="light">
					<MagicSelect options={options} allowClear />
				</MagicThemeProvider>,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
