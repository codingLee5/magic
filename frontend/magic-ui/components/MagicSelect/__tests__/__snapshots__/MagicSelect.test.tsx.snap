// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicSelect > 快照测试 > 基础选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-outlined css-var-r4 magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-single magic-select-show-arrow"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <span
          class="magic-select-selection-search"
        >
          <input
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="magic-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="magic-select-selection-placeholder"
        >
          请选择
        </span>
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`MagicSelect > 快照测试 > 多选选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-outlined css-var-ra magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-multiple magic-select-show-arrow magic-select-show-search"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <div
          class="magic-select-selection-overflow"
        >
          <div
            class="magic-select-selection-overflow-item magic-select-selection-overflow-item-suffix"
            style="opacity: 1;"
          >
            <div
              class="magic-select-selection-search"
              style="width: 0px;"
            >
              <input
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autocomplete="off"
                class="magic-select-selection-search-input"
                id="rc_select_TEST_OR_SSR"
                readonly=""
                role="combobox"
                style="opacity: 0;"
                type="search"
                unselectable="on"
                value=""
              />
              <span
                aria-hidden="true"
                class="magic-select-selection-search-mirror"
              >
                 
              </span>
            </div>
          </div>
        </div>
        <span
          class="magic-select-selection-placeholder"
        />
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`MagicSelect > 快照测试 > 大尺寸选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-lg magic-select-outlined css-var-rc magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-single magic-select-show-arrow"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <span
          class="magic-select-selection-search"
        >
          <input
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="magic-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="magic-select-selection-placeholder"
        />
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`MagicSelect > 快照测试 > 小尺寸选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-sm magic-select-outlined css-var-re magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-single magic-select-show-arrow"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <span
          class="magic-select-selection-search"
        >
          <input
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="magic-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="magic-select-selection-placeholder"
        />
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`MagicSelect > 快照测试 > 带清除按钮选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-outlined css-var-rg magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-single magic-select-allow-clear magic-select-show-arrow"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <span
          class="magic-select-selection-search"
        >
          <input
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="magic-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="magic-select-selection-placeholder"
        />
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`MagicSelect > 快照测试 > 带默认值选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-outlined css-var-r6 magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-single magic-select-show-arrow"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <span
          class="magic-select-selection-search"
        >
          <input
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="magic-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="magic-select-selection-item"
          title="选项1"
        >
          选项1
        </span>
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`MagicSelect > 快照测试 > 禁用状态选择器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-select magic-select-outlined css-var-r8 magic-select-css-var css-dev-only-do-not-override-yotw4n magic-select-single magic-select-show-arrow magic-select-disabled"
  >
    <div
      class="magic-select-selector"
    >
      <span
        class="magic-select-selection-wrap"
      >
        <span
          class="magic-select-selection-search"
        >
          <input
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="magic-select-selection-search-input"
            disabled=""
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="magic-select-selection-placeholder"
        />
      </span>
    </div>
    <span
      aria-hidden="true"
      class="magic-select-arrow"
      style="user-select: none;"
      unselectable="on"
    >
      <svg
        class="tabler-icon tabler-icon-chevron-down "
        fill="none"
        height="16"
        stroke="rgba(28,29,35, 0.8)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9l6 6l6 -6"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;
